addons/bestja_stores/models.py

```diff
@@ -266,10 +272,96 @@ def _default_project(self):
     days = fields.One2many('bestja_stores.day', inverse_name='store')
-    state = fields.Selection(STATES, default='waiting_bank', required=True, string="Status")
+    state = fields.Selection(STATES, default='waiting_bank', required=True, string="Status aktywacji")
+    time_deactivated = fields.Datetime(string="Czas dezaktywacji")
     name = fields.Char(related='store.name', readonly=True)
     address = fields.Char(related='store.address', readonly=True)
     city = fields.Char(related='store.city', readonly=True)
+    user_can_moderate = fields.Boolean(compute="_compute_user_can_moderate")
+
+    @api.model
+    def _needaction_domain_get(self):
+        """
+        Show pending count in menu.
+        """
+        return [
+            ('state', 'in', ['waiting_bank', 'waiting_partner']),
+        ]
+
+    @api.multi
+    def _is_permitted(self):
+        """
+        Allow authorized users to modify protected fields
+        """
+        permitted = super(StoreInProject, self)._is_permitted()
+        return permitted or (self.user_can_moderate and self.state != 'activated')
+
+    @api.multi
+    def is_bank(self):
+        """
+        Does the current user have privileges to act as bank
+        (organization level 1) overseeing current project?
+        """
+        self.ensure_one()
+        authorised_uids = [
+            self.project.parent.organization.coordinator.id,
+            self.project.parent.manager.id,
+        ]
+        return self.env.uid in authorised_uids and self.project.organization_level == 2
+
+    @api.multi
+    def is_owner(self):
+        """
+        Does the current user have privileges to act as an owner
+        of the current project.
+        """
+        self.ensure_one()
+        authorised_uids = [
+            self.project.organization.coordinator.id,
+            self.project.manager.id,
+        ]
+        return self.env.uid in authorised_uids
+
+    @api.one
+    @api.depends('project', 'project.parent')
+    def _compute_user_can_moderate(self):
+        """
+        Is current user authorized to moderate (accept/reject) the store?
+        """
+        if self.state == 'waiting_bank':
+            self.user_can_moderate = self.is_bank()
+        elif self.state in ['waiting_partner', 'activated']:
+            self.user_can_moderate = self.is_owner() or self.is_bank()
+        else:
+            self.user_can_moderate = False
+
+    @api.one
+    def set_activated(self):
+        self.state = 'activated'
+        if self.is_bank():
+            self.activated_by = self.project.parent.organization
+        elif self.is_owner():
+            self.activated_by = self.project.organization
+        else:
+            # This shouldn't really happen, but we can't forbid super user
+            # from doing anything, so theoretically speaking it might...
+            self.activated_by = False
+
+    @api.one
+    def set_deactivated(self):
+        if not (self.is_owner() or self.is_bank()):
```

Comment:
Why do we check permissions here, but we don’t check them in `set_activated` and `set_rejected` methods?